### ISSUE TITLE
Use FIFO queue for the slow query

### ISSUE THREAD

**Issue by BusyJay** [tikv/tikv#12188]
### Is your feature request related to a problem? Please describe:
We use unified thread pool to handle coprocessor requests. If the requests take too long to finish, it will be scheduled to the low priority queue. Unified thread pool will pick a task from the low priority queue, execute a time slice, and then reschedule it to the back of the queue.

If a task takes at least 30 seconds to finish, and there are 16 tasks being scheduled, while only 7 threads to handle them, then non of them will get finished. Supposing the scheduling opportunity is fair, then each task will run 29 seconds at the same time, hence each thread has been running for 29 * 16 / 7 = 66.28 seconds > 60 seconds, so all tasks will be cancelled with error deadline exceeded.

### Describe the feature you'd like:

Instead of pushing the long running task to the back of queue, it should be push to the front of the queue. So in the above case, at least (60 / 30) * 7 = 14 tasks can finish.

FIFO can also improve olap jobs probably. On one hand, less failure tasks mean less retry, so it hardly cause avalanche breakdown. On the other hand, most tasks finish in timely fashion, so if the key tasks happens to, and very likely to, finish early, the whole batch processing can finish early.

Instead of using the low priority queue, we should use a separate queue for slow query specifically. Using a separate FIFO queue has an advantage that we can predict whether a slow task can finish when it's enqueued. If the queue becomes too long, we can just discard the slow task while still be able to executing the quick task. So the hard limit on all running tasks can be removed.

### Describe alternatives you've considered:
None.

### Teachability, Documentation, Adoption, Migration Strategy:
None.

**Comment by sticnarf** [comment]
> If a task takes at least 30 seconds to finish, and there are 16 tasks being scheduled, while only 7 threads to handle them, then non of them will get finished.

If it's exactly the case you talk about here, ISTM the semaphore that limits the number of concurrent long tasks should work. Long tasks that fail to acquire permits have to wait until the others release the permits, making these long tasks run just like in a queue.

**Comment by BusyJay** [comment]
Is the high priority tasks share the same semaphore? Or are they bypass the check? If they are using the same semaphore, then high priority may not get expected amount of time slice. If they by pass the check, then it will suffer the similar problem described above.

**Comment by sticnarf** [comment]
Only the tasks that take more than 5 milliseconds may be blocked by the semaphore. In this way, we try to avoid the problem described in this issue while letting short tasks get enough time slice.

**Comment by BusyJay** [comment]
Then there still be two improvements missing.

1. Different priority tasks may use different semaphores.
2. There should be an additional limit work on the pending tasks waiting for semaphores.
